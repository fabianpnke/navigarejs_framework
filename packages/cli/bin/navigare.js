#!/usr/bin/env node
import cli from './../build/index.modern.js'

cli.parse(process.argv)
