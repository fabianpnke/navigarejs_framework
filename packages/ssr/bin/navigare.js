#!/usr/bin/env node

import { serve } from './../build/index.esm.js'
import path from 'node:path'
;(async () => {
  const input = process.argv[2]
  const finalInput = input ? path.join(process.cwd(), input) : undefined

  const { printUrls } = await serve({
    input: finalInput,
    logger: {
      info: (message) => {
        console.info(message)
      },
      warn: (message) => {
        console.warn(message)
      },
      warnOnce: (message) => {
        console.warn(message)
      },
      error: (message) => {
        console.error(message)
      },
    },
  })

  printUrls()
})()
