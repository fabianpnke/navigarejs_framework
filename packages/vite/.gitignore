node_modules
build
types
tests/dist