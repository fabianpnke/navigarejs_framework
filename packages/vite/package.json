{
  "name": "@navigare/vite",
  "version": "1.0.14",
  "license": "MIT",
  "type": "commonjs",
  "publishConfig": {
    "access": "public"
  },
  "description": "A framework for creating server-driven single page apps.",
  "keywords": [
    "vite-plugin"
  ],
  "contributors": [
    "Julian Hundeloh <github@julian.fyi>"
  ],
  "funding": [
    {
      "type": "github",
      "url": "https://github.com/sponsors/jaulz"
    }
  ],
  "homepage": "https://navigarejs.github.io/framework/",
  "repository": {
    "type": "git",
    "url": "https://github.com/navigarejs/framework.git",
    "directory": "packages/vite"
  },
  "bugs": {
    "url": "https://github.com/navigarejs/framework/issues"
  },
  "files": [
    "build",
    "types"
  ],
  "source": "./src/index.ts",
  "main": "./build/index.js",
  "types": "./types/index.d.ts",
  "scripts": {
    "prepare": "yarn ../core build && yarn ../ssr build",
    "build": "yarn run clean && tsc -p tsconfig.build.json",
    "clean": "rm -rf types && rm -rf build",
    "dev": "tsc-watch -p tsconfig.build.json",
    "test": "chmod 777 ./tests/__fixtures__/laravel/artisan.js && vitest",
    "prepack": "yarn ../core prepack > /dev/null && yarn ../ssr prepack > /dev/null && yarn prepack > /dev/null",
    "release": "yarn semantic-release"
  },
  "devDependencies": {
    "@navigare/core": "1.0.18",
    "@navigare/ssr": "1.0.14",
    "@semantic-release/changelog": "^6.0.1",
    "@semantic-release/exec": "^6.0.3",
    "@trivago/prettier-plugin-sort-imports": "^3.4.0",
    "@types/babel__generator": "^7.6.4",
    "@types/babel__template": "^7.4.1",
    "@types/babel__traverse": "^7.18.2",
    "@types/debug": "^4.1.7",
    "@types/lodash.defaultsdeep": "^4.6.7",
    "@types/node": "^18.11.5",
    "@types/react": "^18.0.22",
    "@types/semantic-release": "^17.2.4",
    "@typescript-eslint/eslint-plugin": "^5.41.0",
    "@typescript-eslint/parser": "^5.41.0",
    "@vitejs/plugin-vue-jsx": "^2.0.1",
    "eslint": "^8.26.0",
    "prettier": "^2.7.1",
    "react": "^18.2.0",
    "rollup": "^2.79.1",
    "semantic-release": "^19.0.5",
    "semantic-release-monorepo": "^7.0.5",
    "tsc-watch": "^5.0.3",
    "typescript": "^4.8.4",
    "vite": "^3.1.8",
    "vitest": "^0.24.3",
    "vue": "^3.2.41"
  },
  "peerDependencies": {
    "@navigare/core": "1.0.15",
    "vite": "^3.0.0"
  },
  "dependencies": {
    "@babel/generator": "^7.19.6",
    "@babel/parser": "^7.19.6",
    "@babel/template": "^7.18.10",
    "@babel/traverse": "^7.19.3",
    "@babel/types": "^7.19.3",
    "@types/lodash.isarray": "^4.0.7",
    "@types/lodash.isobject": "^3.0.7",
    "@types/lodash.isstring": "^4.0.7",
    "chalk": "^4.1.2",
    "debug": "^4.3.4",
    "execa": "^5.1.1",
    "lodash.defaultsdeep": "^4.6.1",
    "lodash.isarray": "^4.0.0",
    "lodash.isobject": "^3.0.2",
    "lodash.isstring": "^4.0.1"
  }
}
