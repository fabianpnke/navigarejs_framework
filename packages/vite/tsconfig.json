{
  "extends": "./../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./build",

    "lib": ["ES2020"],
    "types": ["node"],
    "jsx": "react",

    "declaration": true,
    "declarationDir": "types",

    "typeRoots": ["./node_modules/@types"]
  }
}
