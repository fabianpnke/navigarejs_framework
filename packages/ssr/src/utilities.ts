import { isDefined } from '@navigare/core'
import castArray from 'lodash.castarray'
import { Manifest, ManifestChunk, ModuleNode, ViteDevServer } from 'vite'

export const renderHead = (
  modules: Set<string>,
  manifest: Manifest,
  base: string,
): string[] => {
  const tags: string[] = []
  const visitedIds = new Set<string>()
  const visitedFiles = new Set<string>()

  for (const id of Array.from(modules)) {
    const chunk = manifest[id]

    if (!chunk) {
      continue
    }

    tags.push(
      ...renderHeadChunk(manifest, base, id, chunk, visitedIds, visitedFiles),
    )
  }

  return tags
}

export const renderHeadChunk = (
  manifest: Manifest,
  base: string,
  id: string,
  chunk: ManifestChunk,
  visitedIds: Set<string>,
  visitedFiles: Set<string>,
): string[] => {
  const tags: string[] = []

  if (visitedIds.has(id)) {
    return tags
  }

  // Remember that we already visited this module so we do not end in an infinite loop
  visitedIds.add(id)

  // Collect files plus styles
  const files = [
    ...castArray(chunk.file),
    ...castArray(chunk.css),
    // ...castArray(chunk.dynamicImports),
  ].filter(isDefined)

  // Go through files and render preload links
  for (const file of files) {
    if (visitedFiles.has(file)) {
      continue
    }

    visitedFiles.add(file)

    const preloadLink = generatePreloadLink(file, base)
    if (!preloadLink) {
      continue
    }

    tags.push(preloadLink)
  }

  // Collect dynamic imports
  const dynamicTags = castArray(chunk.dynamicImports)
    .filter(isDefined)
    .map((dynamicImportId) => {
      const dynamicImportChunk = manifest[dynamicImportId]

      return renderHeadChunk(
        manifest,
        base,
        dynamicImportId,
        dynamicImportChunk,
        visitedIds,
        visitedFiles,
      )
    })
    .flat()

  tags.push(...dynamicTags)

  return tags
}

export const generatePreloadLink = (
  _file: string,
  _base: string,
): string | undefined => {
  /*const finalFile = `${base}${file}`

  if (finalFile.endsWith('js') || finalFile.endsWith('.vue')) {
    return `<link rel="modulepreload" crossorigin href="${finalFile}">`
  }

  if (finalFile.endsWith('.css')) {
    return `<link rel="stylesheet" href="${finalFile}">`
  } else if (finalFile.endsWith('.woff')) {
    return ` <link rel="preload" href="${finalFile}" as="font" type="font/woff" crossorigin>`
  } else if (finalFile.endsWith('.woff2')) {
    return ` <link rel="preload" href="${finalFile}" as="font" type="font/woff2" crossorigin>`
  } else if (finalFile.endsWith('.gif')) {
    return ` <link rel="preload" href="${finalFile}" as="image" type="image/gif">`
  } else if (finalFile.endsWith('.jpg') || finalFile.endsWith('.jpeg')) {
    return ` <link rel="preload" href="${finalFile}" as="image" type="image/jpeg">`
  } else if (finalFile.endsWith('.png')) {
    return ` <link rel="preload" href="${finalFile}" as="image" type="image/png">`
  }*/

  return undefined
}

export const getStylesFromFiles = async (
  vite: ViteDevServer | undefined,
  files: Set<string>,
): Promise<string> => {
  if (!vite) {
    return ''
  }

  const styles: string[] = []
  for (const file of files) {
    const modules = await getModulesByFile(vite, file)

    if (!modules) {
      continue
    }

    styles.push(...getStylesFromModules(modules))
  }

  return styles.join('\n\n')
}

export const getModulesByFile = async (
  vite: ViteDevServer,
  component: string,
) => {
  const filePath = '/Users/julian/Development/projects/empacity/' + component

  return await vite.moduleGraph.getModulesByFile(filePath)
}

export const getStylesFromModules = (modules: Set<ModuleNode>) => {
  const styles: string[] = []

  for (const module of modules) {
    if (
      !module.id?.endsWith('.css') &&
      !module.id?.includes('vue&type=style')
    ) {
      continue
    }

    styles.push(module.ssrModule?.default)
  }

  return styles
}
