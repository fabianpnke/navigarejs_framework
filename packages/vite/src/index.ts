import { getConfiguration } from './configuration'
import { getRoutes, writeTypes } from './routes'
import { Options, Adapter, Configuration, Plugin } from './types'
import generate from '@babel/generator'
import { parse } from '@babel/parser'
import traverse, { NodePath } from '@babel/traverse'
import {
  isImportSpecifier,
  isIdentifier,
  isCallExpression,
  identifier,
  isStringLiteral,
  isMemberExpression,
  CallExpression,
  callExpression,
  Expression,
  ArgumentPlaceholder,
  JSXNamespacedName,
  SpreadElement,
  Program,
  importDeclaration,
  importSpecifier,
  stringLiteral,
} from '@babel/types'
import { RawRoutes, throwError } from '@navigare/core'
import { Server as SSRServer, serveSSR } from '@navigare/ssr'
import makeDebugger from 'debug'
import defaultsDeep from 'lodash.defaultsdeep'
import { createRequire } from 'module'
import path from 'path'
import { createLogger, loadEnv, Logger } from 'vite'

const require = createRequire(import.meta.url)

const debug = makeDebugger('navigare:laravel:index')

export default function createNavigarePlugin(options: Options = {}): Plugin {
  defaultsDeep(options, {
    routes: Adapter.Laravel,
    configuration: Adapter.Laravel,
    interval: 15000,
  })
  let currentRoutes: RawRoutes | null = null
  let currentConfiguration: Configuration | null = null
  let ssr = false
  let ssrServer: SSRServer | undefined = undefined
  let logger: Logger | undefined = undefined

  return {
    name: 'navigare',

    enforce: 'post',

    buildEnd() {
      // Nothing yet
    },

    closeBundle() {
      // Nothing yet
    },

    async config(configuration, { mode, command }) {
      const serving = command === 'serve'

      // Create logger
      logger = createLogger(configuration.logLevel, {
        allowClearScreen: configuration.clearScreen,
        customLogger: configuration.customLogger,
        prefix: 'navigare',
      })

      // Enable manifest
      configuration.build = {
        ...configuration.build,
        manifest: true,
      }

      // Set SSR mode
      ssr = !!configuration.build?.ssr
      debug('set SSR mode to %s', ssr)

      // Update SSR config
      if (ssr) {
        configuration.ssr = configuration.ssr || {}
        configuration.ssr.external = [
          ...(configuration.ssr?.external || []),

          ...(serving
            ? [
                // Core
                ...[
                  'lodash.clonedeep',
                  'lodash.debounce',
                  'lodash.get',
                  'lodash.isarray',
                  'lodash.isequal',
                  'lodash.isboolean',
                  'lodash.isfunction',
                  'lodash.isnumber',
                  'lodash.isobject',
                  'lodash.isstring',
                  'lodash.issymbol',
                  'lodash.merge',
                  'lodash.uniq',
                ],
                // Vue3
                ...['vue', 'lodash.isequal'],
              ]
            : []),
        ]
      }

      // Loads .env
      const env = loadEnv(mode, process.cwd(), '')

      // Read routes regularly in case no routes were provided
      const updateRoutes = async () => {
        try {
          currentRoutes = await getRoutes(options, env)
          debug('read routes: %O', currentRoutes)

          // Write types
          if (currentRoutes) {
            // await writeTypes(currentConfiguration?.types.path, currentRoutes)
            await writeTypes(
              path.join(
                require.resolve('@navigare/core'),
                '..',
                '..',
                'routes.d.ts',
              ),
              currentRoutes,
            )
          }
        } catch (error) {
          if (error instanceof Error) {
            console.error(error.message)
          }
        }

        // Update routes in a while again
        const interval = Number(options.interval)
        if (serving && interval > 0) {
          setTimeout(updateRoutes, interval)
        }
      }
      await updateRoutes()

      // Read configuration regularly in case no routes were provided
      const updateConfiguration = async () => {
        try {
          currentConfiguration = await getConfiguration(options, env)
          debug('read configuration: %O', currentConfiguration)
        } catch (error) {
          if (error instanceof Error) {
            console.error(error.message)
          }
        }

        // Update configuration in a while again
        const interval = Number(options.interval)
        if (serving && interval > 0) {
          setTimeout(updateConfiguration, interval)
        }
      }
      await updateConfiguration()
    },

    async configureServer(server) {
      // In case we are already in SSR mode, we don't need to start the server
      if (ssr) {
        // Catch uncaught exceptions
        process.on('uncaughtException', (error) => {
          logger?.error(
            `We observed an uncaught exception. Did you try to access a global variable like "window" which is not available in Node environment?
    ${error.stack}`,
            {
              clear: true,
              timestamp: true,
              error: error,
            },
          )
        })

        return
      }

      // Start SSR server
      debug('serving SSR at %s', currentConfiguration?.ssr.port)
      ssrServer = await serveSSR({
        logger,
        port: currentConfiguration?.ssr.port,
      })

      server.httpServer?.once('listening', () => {
        setTimeout(() => {
          ssrServer?.printUrls()
        }, 25)
      })
    },

    async transform(code: string, id: string) {
      if (!code.includes('route')) {
        return
      }

      if (id.includes('node_modules')) {
        return
      }

      // Parse code
      const ast = parse(code, {
        sourceType: 'module',
        plugins: ['jsx', 'typescript'],
      })

      let changed = false
      const replaceArgumentWithRoute = (routeCall: CallExpression) => {
        if (!isStringLiteral(routeCall.arguments[0])) {
          /*console.log(routeCall)
          throwError(
            'the first parameter of "route" function must be a string literal',
            routeCall.arguments,
          )*/
          return
        }

        const routeName = routeCall.arguments[0].value
        const route = currentRoutes?.[routeName]
        if (!route) {
          throwError(`"${routeName}" is not a valid route name`)
        }

        routeCall.arguments[0] = identifier(
          JSON.stringify(currentRoutes![routeCall.arguments[0].value]),
        )
        changed = true
      }
      const replaceWithRouteCall = (
        program: NodePath<Program>,
        path: NodePath,
        args: (
          | Expression
          | ArgumentPlaceholder
          | JSXNamespacedName
          | SpreadElement
        )[],
      ) => {
        const routeIdentifier = program.scope.generateUidIdentifier('route')
        const [importDeclarationPath] = program.unshiftContainer(
          'body',
          importDeclaration(
            [importSpecifier(routeIdentifier, identifier('route'))],
            stringLiteral('@navigare/core'),
          ),
        )
        program.scope.registerDeclaration(importDeclarationPath)

        const binding = program.scope.getBinding(routeIdentifier.name)
        const [routeCallExpressionPath] = path.replaceWith(
          callExpression(routeIdentifier, args),
        )
        binding?.reference(routeCallExpressionPath.get('callee'))

        // Remember that we changed something, so we can output the new formatted AST later
        changed = true
      }

      traverse(ast, {
        Program(path) {
          const program = path

          // Turn special cases into route(...) calls:
          // - _ctx.route(...)
          // - $setup.route(...)
          // - __unref(route)(...)
          const objectIdentifierNames = ['_ctx', '$setup']
          const calleeIdentifierNames = ['_unref']
          const propertyIdentifierNames = ['route']

          traverse(ast, {
            CallExpression(path) {
              const { callee } = path.node

              // Member expressions like _ctx.route(...) or $setup.route(...)
              if (isMemberExpression(callee)) {
                if (!isIdentifier(callee.object)) {
                  return
                }

                if (!isIdentifier(callee.property)) {
                  return
                }

                if (!objectIdentifierNames.includes(callee.object.name)) {
                  return
                }

                if (!propertyIdentifierNames.includes(callee.property.name)) {
                  return
                }

                replaceWithRouteCall(program, path, path.node.arguments)
                return
              }

              // Call expressions like __unref(route)(...)
              if (isCallExpression(callee)) {
                if (!isIdentifier(callee.callee)) {
                  return
                }

                if (!calleeIdentifierNames.includes(callee.callee.name)) {
                  return
                }

                if (!isIdentifier(callee.arguments[0])) {
                  return
                }

                if (
                  !propertyIdentifierNames.includes(callee.arguments[0].name)
                ) {
                  return
                }

                replaceWithRouteCall(program, path, path.node.arguments)
                return
              }
            },
          })

          // Catch function calls like route(...)
          traverse(ast, {
            ImportDeclaration(path) {
              const { specifiers, source } = path.node

              if (source.value !== '@navigare/core') {
                return
              }

              // Check if the route function was imported
              const routeIdentifiers = specifiers
                .map((specifier) => {
                  if (!isImportSpecifier(specifier)) {
                    return null
                  }

                  if (!isIdentifier(specifier.imported)) {
                    return null
                  }

                  if (specifier.imported.name !== 'route') {
                    return null
                  }

                  return specifier.local.name
                })
                .filter((identifier): identifier is string => !!identifier)

              for (const routeIdentifier of routeIdentifiers) {
                const binding = program.scope?.getBinding(routeIdentifier)
                for (const reference of binding?.referencePaths ?? []) {
                  if (!isCallExpression(reference.parent)) {
                    return
                  }

                  replaceArgumentWithRoute(reference.parent)
                }
              }
            },
          })
        },
      })

      if (!changed) {
        return
      }

      // Return updated code
      const updatedCode = generate(ast, {}, code).code

      return updatedCode
    },
  }
}
