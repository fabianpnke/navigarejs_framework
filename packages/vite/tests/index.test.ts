import navigare from '../src'
import routes from './__fixtures__/routes.json'
import { fixture } from './utilities'
import vue from '@vitejs/plugin-vue'
import vueJSX from '@vitejs/plugin-vue-jsx'
import path from 'node:path'
import { build as vite } from 'vite'
import { describe, expect, it } from 'vitest'

/*
const copy = (value: string) => {
  const proc = require('child_process').spawn('pbcopy')
  proc.stdin.write(value)
  proc.stdin.end()
}
*/

const build = async (input: string) => {
  const result = (await vite({
    mode: 'production',
    root: path.resolve(__dirname),
    build: {
      // outDir: fixture('test/'),
      minify: false,
      rollupOptions: {
        input,
        external: [/node_modules/, /@navigare/],
      },
    },
    logLevel: 'silent',
    plugins: [
      vue({}),

      vueJSX(),

      navigare({
        interval: false,
        configuration: fixture('configuration.json'),
        routes: fixture('routes.json'),
      }),
    ],
  })) as any

  return result.output.reduce(
    (
      cumulatedCode: string,
      output: {
        code: string
      },
    ) => {
      return [cumulatedCode, output.code].join('\n')
    },
    '',
  )
}

it('throw an error for invalid route names', async () => {
  try {
    await build(fixture('resources/scripts/invalid.ts'))
  } catch (error: any) {
    expect(error.message).toBe('Navigare: "invalid" is not a valid route name.')
  }
})

const findRouteInCode = (route: any, code: string): number => {
  const codeWithoutWhitespaces = code.replace(/\s/g, '')

  /*return (
    codeWithoutWhitespaces
      .replace(/\s/g, '')
      .split(JSON.stringify(route).replace(/\s/g, '')).length - 1
  )*/

  return codeWithoutWhitespaces.split(route.uri).length - 1
}

describe('Vue', () => {
  it('replaces routes in Vue components (*.ts)', async () => {
    const code = await build(
      fixture('resources/scripts/vue/ComponentContainer.ts'),
    )

    expect(code).toMatchSnapshot()
    expect(findRouteInCode(routes.welcome, code)).toEqual(1)
  })

  it('replaces routes in Vue components (*.vue)', async () => {
    const code = await build(
      fixture('resources/scripts/vue/ComponentContainer.vue'),
    )

    expect(code).toMatchSnapshot()
    expect(findRouteInCode(routes.welcome, code)).toEqual(2)
  })

  it('replaces routes in Vue components with setup sugar', async () => {
    const code = await build(
      fixture('resources/scripts/vue/SetupContainer.vue'),
    )

    expect(code).toMatchSnapshot()
    expect(findRouteInCode(routes.welcome, code)).toEqual(2)
  })

  it('replaces routes in Vue components with TSX', async () => {
    const code = await build(fixture('resources/scripts/vue/Component.tsx'))

    expect(code).toMatchSnapshot()
    expect(findRouteInCode(routes.welcome, code)).toEqual(2)
  })
})

describe.skip('React', () => {
  it('replaces routes in React components with TSX', async () => {
    const code = await build(fixture('resources/scripts/react/Component.tsx'))

    expect(code).toMatchSnapshot()
    expect(findRouteInCode(routes.welcome, code)).toEqual(2)
  })
})
