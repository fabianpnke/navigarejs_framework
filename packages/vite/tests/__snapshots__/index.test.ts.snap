// Vitest Snapshot v1

exports[`Vue > replaces routes in Vue components (*.ts) 1`] = `
"
import { defineComponent } from \\"../../../../../../node_modules/vue/dist/vue.runtime.esm-bundler.js\\";
import { route } from \\"@navigare/core\\";
const Component = defineComponent({
  setup() {
    const rawRoute = route({ \\"name\\": \\"welcome\\", \\"uri\\": \\"welcome/{name?}\\", \\"method\\": \\"GET\\", \\"methods\\": [\\"GET\\", \\"HEAD\\"], \\"domain\\": \\"navigare.test\\", \\"bindings\\": { \\"name\\": null }, \\"components\\": [{ \\"id\\": \\"nested/Details.vue\\", \\"path\\": \\"1fa1dabf45da756f169804234fdb296c.mjs\\" }] });
    return {
      rawRoute
    };
  },
  template: \`
    <div>Hello World</div>
    <div>{{rawRoute}}</div>
    <div>{{route('welcome')}}</div>
  \`
});
defineComponent({
  components: {
    Component
  },
  template: \`
    <component />
  \`
});

"
`;

exports[`Vue > replaces routes in Vue components (*.vue) 1`] = `
"
import { defineComponent, openBlock, createElementBlock, Fragment, createElementVNode, toDisplayString } from \\"../../../../../../node_modules/vue/dist/vue.runtime.esm-bundler.js\\";
import { route } from \\"@navigare/core\\";
const _sfc_main = defineComponent({
  setup() {
    const rawRoute = route({ \\"name\\": \\"welcome\\", \\"uri\\": \\"welcome/{name?}\\", \\"method\\": \\"GET\\", \\"methods\\": [\\"GET\\", \\"HEAD\\"], \\"domain\\": \\"navigare.test\\", \\"bindings\\": { \\"name\\": null }, \\"components\\": [{ \\"id\\": \\"nested/Details.vue\\", \\"path\\": \\"29bbb271369c636446f8e9f39e9d9a41.mjs\\" }] });
    return {
      rawRoute
    };
  }
});
const _export_sfc = (sfc, props) => {
  const target = sfc.__vccOpts || sfc;
  for (const [key, val] of props) {
    target[key] = val;
  }
  return target;
};
const _hoisted_1 = /* @__PURE__ */ createElementVNode(\\"div\\", null, \\"Hello World\\", -1);
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock(Fragment, null, [_hoisted_1, createElementVNode(\\"div\\", null, toDisplayString(_ctx.rawRoute), 1), createElementVNode(\\"div\\", null, toDisplayString(route({ \\"name\\": \\"welcome\\", \\"uri\\": \\"welcome/{name?}\\", \\"method\\": \\"GET\\", \\"methods\\": [\\"GET\\", \\"HEAD\\"], \\"domain\\": \\"navigare.test\\", \\"bindings\\": { \\"name\\": null }, \\"components\\": [{ \\"id\\": \\"nested/Details.vue\\", \\"path\\": \\"29bbb271369c636446f8e9f39e9d9a41.mjs\\" }] })), 1)], 64);
}
const Component = /* @__PURE__ */ _export_sfc(_sfc_main, [[\\"render\\", _sfc_render]]);
defineComponent({
  components: {
    Component
  }
});

"
`;

exports[`Vue > replaces routes in Vue components with TSX 1`] = `
"
import { defineComponent, createVNode, Fragment, createTextVNode } from \\"../../../../../../node_modules/vue/dist/vue.runtime.esm-bundler.js\\";
import { route } from \\"@navigare/core\\";
defineComponent({
  name: \\"Component\\",
  setup() {
    const rawRoute = route({ \\"name\\": \\"welcome\\", \\"uri\\": \\"welcome/{name?}\\", \\"method\\": \\"GET\\", \\"methods\\": [\\"GET\\", \\"HEAD\\"], \\"domain\\": \\"navigare.test\\", \\"bindings\\": { \\"name\\": null }, \\"components\\": [{ \\"id\\": \\"nested/Details.vue\\", \\"path\\": \\"ca61ce9b117ec36665e93e481d19cd2a.mjs\\" }] });
    return {
      rawRoute
    };
  },
  render(props) {
    return createVNode(Fragment, null, [createVNode(\\"div\\", null, [createTextVNode(\\"Hello World\\")]), createVNode(\\"div\\", null, [props.rawRoute]), createVNode(\\"div\\", null, [route({ \\"name\\": \\"welcome\\", \\"uri\\": \\"welcome/{name?}\\", \\"method\\": \\"GET\\", \\"methods\\": [\\"GET\\", \\"HEAD\\"], \\"domain\\": \\"navigare.test\\", \\"bindings\\": { \\"name\\": null }, \\"components\\": [{ \\"id\\": \\"nested/Details.vue\\", \\"path\\": \\"ca61ce9b117ec36665e93e481d19cd2a.mjs\\" }] })])]);
  }
});

"
`;

exports[`Vue > replaces routes in Vue components with setup sugar 1`] = `
"
import { defineComponent, openBlock, createElementBlock, Fragment, createElementVNode, toDisplayString, unref } from \\"../../../../../../node_modules/vue/dist/vue.runtime.esm-bundler.js\\";
import { route } from \\"@navigare/core\\";
const _hoisted_1 = /* @__PURE__ */ createElementVNode(\\"div\\", null, \\"Hello World\\", -1);
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: \\"Setup\\",
  setup(__props) {
    const rawRoute = route({ \\"name\\": \\"welcome\\", \\"uri\\": \\"welcome/{name?}\\", \\"method\\": \\"GET\\", \\"methods\\": [\\"GET\\", \\"HEAD\\"], \\"domain\\": \\"navigare.test\\", \\"bindings\\": { \\"name\\": null }, \\"components\\": [{ \\"id\\": \\"nested/Details.vue\\", \\"path\\": \\"957305c5de9d82432efd3c72dcd1211b.mjs\\" }] });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(Fragment, null, [_hoisted_1, createElementVNode(\\"div\\", null, toDisplayString(route({ \\"name\\": \\"welcome\\", \\"uri\\": \\"welcome/{name?}\\", \\"method\\": \\"GET\\", \\"methods\\": [\\"GET\\", \\"HEAD\\"], \\"domain\\": \\"navigare.test\\", \\"bindings\\": { \\"name\\": null }, \\"components\\": [{ \\"id\\": \\"nested/Details.vue\\", \\"path\\": \\"957305c5de9d82432efd3c72dcd1211b.mjs\\" }] })), 1), createElementVNode(\\"div\\", null, toDisplayString(unref(rawRoute)), 1)], 64);
    };
  }
});
defineComponent({
  components: {
    Setup: _sfc_main
  }
});

"
`;
